django/db/models/sql/subqueries.py

```diff
@@ -82,6 +83,16 @@ def add_update_values(self, values):
         """
         values_seq = []
         for name, val in values.items():
+            if LOOKUP_SEP in name:
+                attributes = name.split(LOOKUP_SEP)
+                name = attributes.pop(0)
+                hierarchy = ','.join(attributes)
+                field = self.get_meta().get_field(name)
+                from django.contrib.postgres.functions import JSONBSet
+                from django.db.models import Value
+                val = JSONBSet(
+                    field.name, Value('{%s}' % hierarchy), Value('"%s"' % val), Value(False)
```

Comment:
Should the val argument should be the following to accommodate primitive values that aren't strings, and that don't necessarily stringify into valid JSON (such as booleans)? Test coverage should likely also be expanded accordingly. We're using the following in our codebase for the third argument to jsonb_set:

`Cast(Value(json.dumps(value)), output_field=JSONField())`